optuna/storages/base.py

```diff
@@ -29,157 +28,539 @@ class BaseStorage(object, metaclass=abc.ABCMeta):
+        Args:
+            study_id:
+                Study ID of a study to read.
+        Returns:
+            Optimization direction of the study.
+
+        Raises:
+            KeyError:
+                If no study with the matching `study_id` exists.
+        """
 
         raise NotImplementedError
 
     @abc.abstractmethod
-    def get_study_user_attrs(self, study_id):
-        # type: (int) -> Dict[str, Any]
+    def get_study_user_attrs(self, study_id: int) -> Dict[str, Any]:
+
+        """Read a dictionary of user-defined attributes of a specified study.
+
+        Args:
+            study_id:
+                Study ID of a study to read.
+        Returns:
+            A dictionary of a user attributes of a study.
+
+        Raises:
+            KeyError:
+                If no study with the matching `study_id` exists.
+        """
 
         raise NotImplementedError
 
     @abc.abstractmethod
-    def get_study_system_attrs(self, study_id):
-        # type: (int) -> Dict[str, Any]
+    def get_study_system_attrs(self, study_id: int) -> Dict[str, Any]:
+
+        """Read a dictionary of optuna-internal attributes of a specified study.
+
+        Args:
+            study_id:
+                Study ID of a study to read.
+        Returns:
+            A dictionary of a system attributes of a study.
+
+        Raises:
+            KeyError:
+                If no study with the matching `study_id` exists.
+        """
 
         raise NotImplementedError
 
     @abc.abstractmethod
-    def get_all_study_summaries(self):
-        # type: () -> List[study.StudySummary]
+    def get_all_study_summaries(self) -> List[study.StudySummary]:
 
-        raise NotImplementedError
+        """Returns a list of `study.StudySummary` objects.
+
+        Returns:
+            A list of `study.StudySummary` objects.
+
+        """
 
     # Basic trial manipulation
 
     @abc.abstractmethod
-    def create_new_trial(self, study_id, template_trial=None):
-        # type: (int, Optional[FrozenTrial]) -> int
+    def create_new_trial(
+        self, study_id: int, template_trial: Optional["FrozenTrial"] = None
+    ) -> int:
+
+        """Create and add a new trial to a specified study.
+
+        Trial ID is unique among all current and deleted trials.
+
+        Args:
+            study_id:
+                Study ID to add a new trial.
+            template_trial:
+                Fronzen trial with default user-attributes, system-attributes,
+                intermediate-values, and a state.
+
+        Returns:
+            Trial ID of the created trial.
+
+        Raises:
+            KeyError:
+                If no study with the matching `study_id` exists.
+        """
 
         raise NotImplementedError
 
     @abc.abstractmethod
-    def set_trial_state(self, trial_id, state):
-        # type: (int, TrialState) -> bool
+    def set_trial_state(self, trial_id: int, state: TrialState) -> bool:
+
+        """Update a state of a specified trial.
```

Comment:
I don't think we should allow the "RUNNING → WAITING" state change. (Both this doc and current implementation allows it now.)

Reply:
This PR won't address this issue.